Add Into<NonNull<T>> impls for Box<T>/Arc<T>/Rc<T>

dwijnand opened the pull request:
/cc @withoutboats who mentioned to me this might be worth adding to the standard library, in withoutboats/smart#4
/cc @kennytm who last year didn't love the idea of having an Into<NonNull<T>> for Box<T>, in #46952

alexcrichton commented:
Thanks for the PR! This seems reasonable to me, but I'm double checking the rest of libstd for other idioms. It looks like this isn't implemented for `Box`, the other obvious candidate. The `Box` type, however, has an `into_raw_non_null` unstable method for doing the same thing (just not via a trait). 

The `Box::into_raw_non_null` method was added in https://github.com/rust-lang/rust/pull/46952 and looks like at least at one point it was `From` but ended up switching back. @SimonSapin do you recall why we opted for not having a `From`/`Into` impl between the `NonNull` and `Box` types?

I think that leaves us with two routes to go here:

* Land these as stable, also removing `Box::into_raw_non_null` in favor of the same pattern here. Note that if we do this I think we'll want to use `From` impls instead of `Into` which are slightly more general.
* Land these as unstable, renaming to `Arc::into_raw_non_null` to mirror `Box`.

I'd ideally like to hear from @SimonSapin first to see if they remember why we didn't do `Into` in the first place.

dwijnand commented:
`From` is not possible because `NonNull` is in libcore and `Arc`/`Rc` is in liballoc, so according to the laws of coherence it can't be implemented.

So the options left are implementing `Into` or implementing it as an inherent method.  (Or option 3, wait for that coherency rule to be lifted via RFC).

SimonSapin commented:
> do you recall why we opted for not having a From/Into impl between the NonNull and Box types?

https://github.com/rust-lang/rust/pull/46952#issuecomment-353956225 mentions that we generally avoid methods that take `self` (rather than associated functions that take `Self`) on generic smart pointers like `Box<T>` because it could shadow a method of the same name on `T`.

Perhaps more importantly, https://github.com/rust-lang/rust/pull/46952#issuecomment-354479436 says I undid that because of a `incoherent_fundamental_impls` warning: https://github.com/rust-lang/rust/issues/46205

`Box` is `#[fundamental]` so adding impls for it of existing traits is a breaking change as far as I understand. https://github.com/rust-lang/rust/pull/56955 proposes adding `#[fundamental]` to `Arc` and `Rc`.

> we'll want to use From impls instead of Into which are slightly more general

IIRC coherence does not allow this, since `NonNull` is defined in libcore but `Arc` and `Rc` in liballoc.

withoutboats commented:
Yes, for box its a potentially breaking change because someone could have implemented `Into<NonNull<MyType>> for Box<MyType>` because its fundamental. We'd need to at least do a crater run.

I personally prefer using standard conversion methods here, especially since we already do that for references, to avoid users having to look up methods with names like `into_raw_non_null`.

alexcrichton commented:
[I don't think coherence blocks this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f2d56a865085e1fabce7ec4ffa81cb07), so @dwijnand want to switch to `From`, add an impl for `Box`, and we can see what the fallout is on crater?

SimonSapin commented:
It does when generics are involved: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cff9ce9a3b1da269f92da22175236596

dwijnand commented:
Yeah...  Over on Zulip I've been pointed to https://github.com/rust-lang/rfcs/blob/master/text/2451-re-rebalancing-coherence.md, which points to https://github.com/rust-lang/rust/issues/55437 and thus https://github.com/rust-lang/rust/pull/56145.  So if we want to use From and make Box work the same way, we could block this on that PR landing.

SimonSapin commented:
> we generally avoid methods that take `self` (rather than associated functions that take `Self`) on generic smart pointers like `Box<T>` because it could shadow a method of the same name on `T`.

I guess this is the remaining concern with an `Into` impl for `Box` (even through a `From` one). @rust-lang/libs, any thoughts on this?

alexcrichton commented:
Oh oops, my mistake! @dwijnand want to switch to `Into` for `Box` and we can crater?

I'll hold of on formulating whether it's a good idea until we've tested it on crater :)

withoutboats commented:
I'll also note that there's an active lang RFC to change coherence in a way that I think would allow `From` & we are generally inclined to merge, but I also think its backward compatible to switch the Into to From

dwijnand commented:
Hey @alexcrichton, sorry I forgot about your request here until now.

I tried implementing Into for Box, but got:

```
error: conflicting implementations of trait `core::convert::Into<core::ptr::NonNull<_>>` for type `boxed::Box<_>`: (E0119)
   --> src/liballoc/boxed.rs:483:1
    |
483 | impl<T: ?Sized> Into<NonNull<T>> for Box<T> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: #[deny(incoherent_fundamental_impls)] on by default
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
    = note: conflicting implementation in crate `core`:
            - impl<T, U> core::convert::Into<U> for T
              where U: core::convert::From<T>;
    = note: downstream crates may implement trait `core::convert::From<boxed::Box<_>>` for type `core::ptr::NonNull<_>`
```

Trying to allow that made my test case fail to compile:

```
   --> src/liballoc/boxed_test.rs:154:52
    |
154 |     let ptr: std::ptr::NonNull<i32> = Box::from(0).into();
    |                                                    ^^^^ the trait `core::convert::From<std::boxed::Box<{integer}>>` is not implemented for `core::ptr::NonNull<i32>`
    |
    = help: the following implementations were found:
              <core::ptr::NonNull<T> as core::convert::From<&'a T>>
              <core::ptr::NonNull<T> as core::convert::From<&'a mut T>>
              <core::ptr::NonNull<T> as core::convert::From<core::ptr::Unique<T>>>
    = note: required because of the requirements on the impl of `core::convert::Into<core::ptr::NonNull<i32>>` for `std::boxed::Box<{integer}>`
```

I then tried something else (either tweaking the impl or teaking the test, I forget), which led me to this confusing situation:

```
error[E0460]: found possibly newer version of crate `alloc` which `std` depends on
   --> src/liballoc/lib.rs:128:1
    |
128 | extern crate std;
    | ^^^^^^^^^^^^^^^^^
    |
    = note: perhaps that crate needs to be recompiled?
    = note: the following crate versions were found:
            crate `alloc`: /d/rust/build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/deps/liballoc-755c123e9f141b95.rlib
            crate `alloc`: /d/rust/build/x86_64-apple-darwin/stage0-sysroot/lib/rustlib/x86_64-apple-darwin/lib/liballoc-755c123e9f141b95.rlib
            crate `std`: /d/rust/build/x86_64-apple-darwin/stage0-sysroot/lib/rustlib/x86_64-apple-darwin/lib/libstd-5a6b2941d398f5d5.dylib
            crate `std`: /d/rust/build/x86_64-apple-darwin/stage0-sysroot/lib/rustlib/x86_64-apple-darwin/lib/libstd-5a6b2941d398f5d5.rlib

error: aborting due to previous error

For more information about this error, try `rustc --explain E0460`.
error: Could not compile `alloc`.

To learn more, run the command again with --verbose.


command did not execute successfully: "/d/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "test" "--target" "x86_64-apple-darwin" "-j" "8" "--release" "--features" "panic-unwind backtrace" "--manifest-path" "/d/rust/src/libstd/Cargo.toml" "-p" "alloc" "--" "to_nonnull" "--quiet"
expected success, got: exit code: 101


failed to run: /d/rust/build/bootstrap/debug/bootstrap test --keep-stage 0 --stage 0 src/liballoc --test-args to_nonnull
Build completed unsuccessfully in 0:00:03
```

So I dropped the `-keep-stage 0` (which I had been using successfully, and had been recommended on Zulip) but that gave me:

```
error[E0460]: found possibly newer version of crate `std` which `getopts` depends on
  --> src/libtest/lib.rs:45:1
   |
45 | extern crate getopts;
   | ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: perhaps that crate needs to be recompiled?
   = note: the following crate versions were found:
           crate `std`: /d/rust/build/x86_64-apple-darwin/stage0-sysroot/lib/rustlib/x86_64-apple-darwin/lib/libstd-5a6b2941d398f5d5.rlib
           crate `std`: /d/rust/build/x86_64-apple-darwin/stage0-sysroot/lib/rustlib/x86_64-apple-darwin/lib/libstd-5a6b2941d398f5d5.dylib
           crate `getopts`: /d/rust/build/x86_64-apple-darwin/stage0-test/x86_64-apple-darwin/release/deps/libgetopts-8c1bca47bf6894d5.rlib

error: aborting due to previous error

For more information about this error, try `rustc --explain E0460`.
error: Could not compile `test`.

To learn more, run the command again with --verbose.
command did not execute successfully: "/d/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "build" "--target" "x86_64-apple-darwin" "-j" "8" "--release" "--manifest-path" "/d/rust/src/libtest/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101
failed to run: /d/rust/build/bootstrap/debug/bootstrap test --stage 0 src/liballoc --test-args to_nonnull
Build completed unsuccessfully in 0:00:17
```

So I dropped my changes entirely and "Building stage0 compiler artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)" took 12m30 (is that normal?) and ended up failing anyways:

```
     Running build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/deps/alloc-4f5d60f9094e66f5

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 100 filtered out

     Running build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/deps/collectionstests-afe1c7c2f5f7578e

running 2 tests
..
test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 560 filtered out

   Doc-tests alloc
error: failed to find a `codegen-backends` folder in the sysroot candidates:
* /d/rust/build/x86_64-apple-darwin/stage0-tools/x86_64-apple-darwin
* /d/rust/build/x86_64-apple-darwin/stage0-sysroot/lib/rustlib/x86_64-apple-darwin
* /d/rust/build/x86_64-apple-darwin/stage0-sysroot

error: test failed, to rerun pass '--doc'


command did not execute successfully: "/d/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "test" "--target" "x86_64-apple-darwin" "-j" "8" "--release" "--features" "panic-unwind backtrace" "--manifest-path" "/d/rust/src/libstd/Cargo.toml" "-p" "alloc" "--" "to_nonnull" "--quiet"
expected success, got: exit code: 1


failed to run: /d/rust/build/bootstrap/debug/bootstrap test --stage 0 src/liballoc --test-args to_nonnull
Build completed unsuccessfully in 0:08:46
```

Actually I just tested any change to the compiler (like an attribute string) and anything leads me to "error[E0460]: found possibly newer version of crate `std` which `getopts` depends on", so I don't think I'll be able to develop this patch any further unless I can unblock my local development of Rust...

SimonSapin commented:
As far as I understand, `found possibly newer version of crate` can occur because of a bug in rustbuild and/or Cargo. One way to get out of that situation is `./x.py clean` or `cargo clean`. (Unfortunately I don’t know a better way.)

As to the initial `incoherent_fundamental_impls` warning/error, this is what I mentioned before in https://github.com/rust-lang/rust/pull/56998#issuecomment-449071538. Currently the language rules around `#[fundamental]` do not allow this impl. https://github.com/rust-lang/rust/pull/56998#issuecomment-449126143 suggests that the language will be tweaked and it might be allowed later, but that has not landed yet.

dwijnand commented:
Is it perhaps related to my usage of `RUSTC_WRAPPER=sccache`?

dwijnand commented:
Anyways, thank you @SimonSapin for unblocking me, I've pushed a commit that seems to work for my test cases.

alexcrichton commented:
@bors: try

bors commented:
:hourglass: Trying commit 98b46ba2132e0fef721c0b5ef6c1933c44df310e with merge 185e0799ac53f1168549e11c4fc64ebf18b4a2c1...

alexcrichton commented:
@bors: retry

kennytm commented:
@craterbot run start=master#ec194646fef1a467073ad74b8b68f6f202cfce97 end=try#185e0799ac53f1168549e11c4fc64ebf18b4a2c1 mode=check-only

dwijnand commented:
Thanks @kennytm.

> Estimated end: 2 days

o_O

dwijnand commented:
Any ideas what happened to that bors try/retry?

alexcrichton commented:
Ok the crater report looks good! In that case I think it's worthwhile to propose to merge this.

@rfcbot fcp merge

@dwijnand for the retry and whatnot we were having bors troubles arounds those times, so no need to worry!

SimonSapin commented:
@alexcrichton So are we really ok with going against the guideline on methods for generic smart pointers type? https://rust-lang-nursery.github.io/api-guidelines/predictability.html#c-smart-ptr If so:

The `#[unstable]` attributes should be changed to `#[stable]` since an `impl` cannot actually be unstable when both the trait and the receiver type are stable. (The attribute doesn’t do anything.)

I’ve added `Box<T>` to the title since it also gains a new impl in the diff.

And, since we’re adding that impl, the `Box::into_raw_non_null` function should be deprecated with a message that says to use `.into()` instead.

sfackler commented:
> And, since we’re adding that impl, the `Box::into_raw_non_null` function should be deprecated with a message that says to use `.into()` instead.

I disagree - explicit methods are more discoverable and avoid type inference issues.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/476469735) ([raw log](https://api.travis-ci.org/v3/job/476469735/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:09b782d5:start=1546882310539222743,finish=1546882411740803713,duration=101201580970
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
travis_time:start:tidy
tidy check
[00:03:27] * 569 error codes
[00:03:27] * highest error code: E0721
[00:03:27] tidy error: /checkout/src/liballoc/boxed.rs:542: malformed stability attribute
[00:03:28] some tidy checks failed
[00:03:28] 
[00:03:28] 
[00:03:28] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:03:28] 
[00:03:28] 
[00:03:28] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:03:28] Build completed unsuccessfully in 0:00:46
[00:03:28] Build completed unsuccessfully in 0:00:46
[00:03:28] make: *** [tidy] Error 1
[00:03:28] Makefile:69: recipe for target 'tidy' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0477af3b
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Mon Jan  7 17:37:08 UTC 2019
---
travis_time:end:237f02cb:start=1546882629339011605,finish=1546882629344753950,duration=5742345
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:16531626
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0f6b9666
travis_time:start:0f6b9666
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:02bb55fe
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/476478095) ([raw log](https://api.travis-ci.org/v3/job/476478095/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:073a2704:start=1546884273519505114,finish=1546884357416357265,duration=83896852151
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:03:51]    Compiling rustc-std-workspace-core v1.0.0 (/checkout/src/tools/rustc-std-workspace-core)
[00:03:52]    Compiling alloc v0.0.0 (/checkout/src/liballoc)
[00:03:52]    Compiling rustc-demangle v0.1.10
[00:03:52]    Compiling panic_abort v0.0.0 (/checkout/src/libpanic_abort)
[00:03:53] error: `#[deprecated]` cannot be used in staged api, use `#[rustc_deprecated]` instead
[00:03:53]    --> src/liballoc/boxed.rs:199:5
[00:03:53]     |
[00:03:53] 199 | /     pub fn into_raw_non_null(b: Box<T>) -> NonNull<T> {
[00:03:53] 200 | |         Box::into_unique(b).into()
[00:03:53]     | |_____^
[00:03:53] 
[00:03:53] 
[00:03:53] error[E0541]: unknown meta item 'issue'
[00:03:53]    --> src/liballoc/boxed.rs:542:45
[00:03:53]     |
[00:03:53] 542 | #[stable(feature = "box_into_raw_non_null", issue = "47336", since = "1.32.0")]
[00:03:53]     |                                             ^^^^^^^^^^^^^^^ expected one of `since`, `note`
[00:03:55] error: aborting due to 2 previous errors
[00:03:55] 
[00:03:55] For more information about this error, try `rustc --explain E0541`.
[00:03:55] error: Could not compile `alloc`.
[00:03:55] error: Could not compile `alloc`.
[00:03:55] 
[00:03:55] To learn more, run the command again with --verbose.
[00:03:55] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:03:55] expected success, got: exit code: 101
[00:03:55] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:03:55] Build completed unsuccessfully in 0:00:38
[00:03:55] Makefile:18: recipe for target 'all' failed
[00:03:55] make: *** [all] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:00ba7030
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Mon Jan  7 18:10:01 UTC 2019
---
travis_time:end:04d42c10:start=1546884602348536868,finish=1546884602353976958,duration=5440090
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0a0a8fcb
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:13b8eb1e
travis_time:start:13b8eb1e
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:034d87c7
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
I'll defer to @withoutboats about making a case for a trait impl and/or inherent methods, I believe they feel more strongly than I

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/476561515) ([raw log](https://api.travis-ci.org/v3/job/476561515/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:1b408abc:start=1546896269282331883,finish=1546896345414854807,duration=76132522924
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:04:08]    Compiling rustc-std-workspace-core v1.0.0 (/checkout/src/tools/rustc-std-workspace-core)
[00:04:10]    Compiling alloc v0.0.0 (/checkout/src/liballoc)
[00:04:10]    Compiling panic_abort v0.0.0 (/checkout/src/libpanic_abort)
[00:04:10]    Compiling rustc-demangle v0.1.10
[00:04:11] error[E0541]: unknown meta item 'note'
[00:04:11]    --> src/liballoc/boxed.rs:198:24
[00:04:11]     |
[00:04:11] 198 |     #[rustc_deprecated(note = "Use `.into()`", since = "1.32.0")]
[00:04:11]     |                        ^^^^^^^^^^^^^^^^^^^^^^ expected one of `since`, `reason`
[00:04:13] error: aborting due to previous error
[00:04:13] 
[00:04:13] For more information about this error, try `rustc --explain E0541`.
[00:04:13] error: Could not compile `alloc`.
[00:04:13] error: Could not compile `alloc`.
[00:04:13] 
[00:04:13] To learn more, run the command again with --verbose.
[00:04:13] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:04:13] expected success, got: exit code: 101
[00:04:13] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:04:13] Build completed unsuccessfully in 0:00:38
[00:04:13] Makefile:18: recipe for target 'all' failed
[00:04:13] make: *** [all] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:1c0b31d4
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Mon Jan  7 21:30:07 UTC 2019
---
travis_time:end:26e88467:start=1546896608271278563,finish=1546896608277859854,duration=6581291
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:08b72162
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:01bbb092
travis_time:start:01bbb092
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:03518b5c
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/476725114) ([raw log](https://api.travis-ci.org/v3/job/476725114/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:03c17c8b:start=1546935829538851020,finish=1546935900580440097,duration=71041589077
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:15:23] 
[01:15:23] running 118 tests
[01:15:49] .iiiii...i.....i..i...i..i.i..i.ii..i.....i..i....i..........iiii..........i...ii...i.......ii.i.i.i 100/118
[01:15:53] ......iii.i.....ii
[01:15:53] 
[01:15:53]  finished in 30.240
[01:15:53] travis_fold:end:test_debuginfo

---
[01:26:08] error: no global memory allocator found but one is required; link to std or add #[global_allocator] to a static item that implements the GlobalAlloc trait.
[01:26:08] 
[01:26:08] 
[01:26:08] running 418 tests
[01:26:27] .................F.................................................................................. 100/418
[01:26:55] .................................................................................................... 300/418
[01:27:07] .................................................................................................... 400/418
[01:27:10] ..................
[01:27:10] failures:
[01:27:10] failures:
[01:27:10] 
[01:27:10] ---- boxed.rs - boxed::Box<T>::into_raw_non_null (line 189) stdout ----
[01:27:10] error: use of deprecated item '<std::boxed::Box<T>>::into_raw_non_null': Use `.into()`
[01:27:10]  --> boxed.rs:194:15
[01:27:10]   |
[01:27:10] 7 |     let ptr = Box::into_raw_non_null(x);
[01:27:10]   |
[01:27:10] note: lint level defined here
[01:27:10]  --> boxed.rs:189:9
[01:27:10]   |
---
[01:27:10] 
[01:27:10] error: test failed, to rerun pass '--doc'
[01:27:10] 
[01:27:10] 
[01:27:10] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "alloc" "--" "--quiet"
[01:27:10] 
[01:27:10] 
[01:27:10] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:27:10] Build completed unsuccessfully in 0:24:06
[01:27:10] Build completed unsuccessfully in 0:24:06
[01:27:10] make: *** [check] Error 1
[01:27:10] Makefile:48: recipe for target 'check' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:36c01f4c
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Tue Jan  8 09:52:20 UTC 2019
---
travis_time:end:1d1bef50:start=1546941141466547234,finish=1546941141524809516,duration=58262282
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:113ca69e
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:1cb0bd00
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

dwijnand commented:
Travis CI is green but is showing yellow/pending.

Re-syncing...

dwijnand commented:
Oh.  Oops, didn't know rfcbot would do that :-/

SimonSapin commented:
@rfcbot fcp merge

dtolnay commented on src/liballoc/boxed.rs:
````diff
@@ -201,6 +203,7 @@ impl<T: ?Sized> Box<T> {
     /// ```
     #[unstable(feature = "box_into_raw_non_null", issue = "47336")]
     #[inline]
+    #[rustc_deprecated(reason = "Use `.into()`", since = "1.32.0")]
````
I would prefer not to deprecated this. Inherent associated functions are more discoverable and avoid needing to guide type inference, as you see in the `into_raw` implementation above.

```rust
Box::into_raw_non_null(b).as_ptr()
```

```rust
let p: NonNull<T> = b.into();
p.as_ptr()
```

dwijnand replied:
@SimonSapin suggested this.  I'm happy either way.

Do we want one implementation to delegate to the other?  And if yes, which way?

SimonSapin replied:
@dtolnay Should every single impl of conversion traits have a corresponding inherent method? If not, what makes this one special?

dwijnand replied:
Perhaps the very fact that the From/Into traits are type conversions makes them special with regards to ergonomics and type inference.  My understanding was that was one of the reasons for having both From _and_ Into, i.e `Foo::from` avoids type inference, while `bar.into()` method-syntax is nicer to use.  Being forced by the coherence rules I think it might be worth keeping the associated function.

dtolnay commented on src/liballoc/boxed.rs:
```diff
@@ -479,6 +482,15 @@ impl From<Box<str>> for Box<[u8]> {
     }
 }
 
+#[allow(incoherent_fundamental_impls)]
```
Please add a comment explaining the implications of this.

dwijnand replied:
I don't know what they are.  I'm happy to push comments if anyone has any suggestions!

Centril replied:
`incoherent_fundamental_impls` is a `C-future-compatibility` warning; those are essentially hard errors that have been temporarily lowered into warnings to give people time to migrate. Tho in this case it doesn't seem clearly specified in the issue what the warning actually does. However, there's a PR up that is actually making it into a hard error https://github.com/rust-lang/rust/pull/49799. If this impl is admitted into stable Rust (i.e. through `#[allow(incoherent_fundamental_impls)]`) then the warning cannot, AFAIK, be made into a hard error. That seems... bad.

cc @nikomatsakis @rust-lang/wg-traits

dtolnay commented on src/liballoc/boxed.rs:
```diff
@@ -479,6 +482,15 @@ impl From<Box<str>> for Box<[u8]> {
 }
 
+#[allow(incoherent_fundamental_impls)]
+#[stable(feature = "box_into_non_null", since = "1.32.0")]
```
This needs to be since 1.33 because the impl will land in nightly first, which is currently 1.33.0-nightly.

Centril replied:
Actually this needs to be 1.34 since master=>beta promotion will happen on the 15th of January and FCP is 10 days... 9 + 10 = 19,  19 > 15.

dtolnay commented on src/liballoc/rc.rs:
```diff
@@ -1179,6 +1179,14 @@ impl<T> From<Vec<T>> for Rc<[T]> {
     }
 }
 
+#[unstable(feature = "rc_into_nonnull", reason = "newly added", issue = "0")]
```
We don't currently support unstable impls so this would be stable immediately. Please change to a `stable` attribute. Same for Arc.

dtolnay commented on src/liballoc/rc.rs:
```diff
@@ -1179,6 +1179,14 @@ impl<T> From<Vec<T>> for Rc<[T]> {
 }
 
+#[unstable(feature = "rc_into_nonnull", reason = "newly added", issue = "0")]
+impl<T: ?Sized> Into<NonNull<T>> for Rc<T> {
```
I would like to add an equivalent inherent function like we have for Box: Rc::into_raw_non_null and Arc::into_raw_non_null.

dwijnand replied:
I'll wait for https://github.com/rust-lang/rust/pull/56998#discussion_r246454126 to be resolved first.

dwijnand commented:
> Yeah... Over on Zulip I've been pointed to [rust-lang/rfcs:text/2451-re-rebalancing-coherence.md@`master`](https://github.com/rust-lang/rfcs/blob/master/text/2451-re-rebalancing-coherence.md), which points to #55437 and thus #56145. So if we want to use From and make Box work the same way, we could block this on that PR landing.

Well that PR landed, so maybe this can be implemented using `From` now?

dwijnand commented:
> maybe this can be implemented using `From` now?

No, it can't.

```rust
impl<T: ?Sized> From<Arc<T>> for NonNull<T>
```

Can't compile in liballoc because `T` is uncovered.  And it can't be implemented with `NonNull` and `From` in libcore because `Arc` isn't available.

dwijnand commented:
@Centril was this triaged at the meeting, or did you change your mind on nominating it?

Centril commented:
@dwijnand It was. The gist of it is that you cannot do `#[allow(incoherent_fundamental_impls)]` for reasons mentioned in https://github.com/rust-lang/rust/pull/56998#discussion_r246474441. If elaboration is necessary, I'll redirect you to @nikomatsakis. :)

dwijnand commented:
I see!  Thanks for the info.  I'll reformulate this proposal as `Arc/Rc::into_raw_non_null` then, and see how that goes.
